src/Microsoft.ML.Data/Prediction/ICalculateFeatureContribution.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.ML.Data;
+
+namespace Microsoft.ML.Trainers
+{
+    /// <summary>
+    /// Interface for mapping input values to corresponding feature contributions.
+    /// This interface is commonly implemented by predictors.
+    /// </summary>
+    [BestFriend]
+    internal interface IFeatureContributionMapper : IPredictor
+    {
+        /// <summary>
+        /// Get a delegate for mapping Contributions to Features.
+        /// Result will contain vector with topN positive contributions(if available) and
+        /// bottomN negative contributions (if available).
+        /// For example linear predictor will have both negative and positive contributions.
+        /// For trees we will not have negative contributions, so bottom param will be ignored.
+        /// If normalization is requested that resulting values will be normalized to [-1, 1].
+        /// </summary>
+        ValueMapper<TSrc, VBuffer<float>> GetFeatureContributionMapper<TSrc, TDst>(int top, int bottom, bool normalize);
+    }
+
+    /// <summary>
+    /// Allows support for feature contribution calculation by model parameters.
+    /// </summary>
+    public interface ICalculateFeatureContribution
+    {
+        FeatureContributionCalculator FeatureContributionCalculator { get; }
+    }
+
+    /// <summary>
+    /// Support for feature contribution calculation.
+    /// </summary>
+    public sealed class FeatureContributionCalculator
```

Comment:
I'm not seeing how this class is used anywhere in the code.

Can it be deleted?

It appears the way this is working is that we have a public API that takes in a `ICalculateFeatureContribution` (the public interface above) object, and then we internally cast that same object to `IFeatureContributionMapper` (the internal interface above that). I don't see where we are using this public class anywhere in our code.

cc @tomfinley @rogancarr

Reply:
tried to remove it, and just use the IFeatureContributionMapper, but it will need the GetFeatureContributionMapper, and the ValueMapper to be public too.  The ValueMapper is internal atm.

---
In reply to: [265147046](https://github.com/dotnet/machinelearning/pull/2914#discussion_r265147046) [](ancestors = 265147046)

Reply:
I was able to remove it, and make the solution build successfully:

https://github.com/eerhardt/machinelearning/commit/cd1ac9f9829ef75ea6e06083d879e919503855fb

Reply:
oh ic, you've removed it completely. I was merging it with the IFeatureContributionMapper. I'll create an issue.